numpy/ma/core.py

```diff
@@ -5079,24 +5002,64 @@ def mean(self, axis=None, dtype=None, out=None):
         1.5
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            with np.errstate(divide='ignore', invalid='ignore'):
+                ret = self._data.mean(axis, dtype, out, **kwargs)
+            zerocounts = self.count(axis, **kwargs) == 0
+
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    out.__setmask__(True if zerocounts else nomask)
```

Comment:
You're right here too, I need to rethink this part. Probably zerocounts should test whether the product of the shape elements at the specified axes is 0.